### PR TITLE
Feature:`sealer debug` to debug nodes/pods/containers

### LINKED ISSUE #523: support subcommand `sealer debug` to debug the nodes and pods
## Issue Description

Type: *feature request*

### Describe what feature you want
The subcommand `sealer debug` can debug the nodes, pods and containers with plenty of useful tools.  This function will help SRE to find the problems in nodes, pods or containers. 

Debug node will create a debug pod in node, and we can attach to the debug pod to debug the node with the tools provided by debug pod. Just like debug node, debug pod/container will create a debug container in pod, and we can can attach to the debug container to debug pod/container with the tools provided by debug container. 
  
When the debugging is done, the generated debug pods, containers will be cleand automatically.

`sealer debug` can also provide some default image for debugging. 
  
Here are some examples of using cases.

```bash
# Debug Node `mynode` and immediately attach to the debug pod.
$ sealer debug node mynode -it
You don't specify an image, it will use the default image: debug:ubuntu
You can use `--image` to specify an image, just like `sealer debug node NODE_NAME --image busybox`.
The debug pod or container id： sealer-pod-debugger-khwbl.mynode.default
root@mynode:~# ls
motd
root@mynode:~#

# In node `mynode`, executing commands without tty. The commands used are debug pod.
$ sealer debug node mynode ls /

# In node `mynode`, executing commands with tty. The commands used are debug pod.
$ sealer debug node mynode -it bash

# Use image `busybox` to debug node `mynode`.
$ sealer debug node mynode -it --image busybox

# It will only a debug pod in node, but will not immediately attach to the debug pod. Users can use the returned debug ID to attach to the debug pod to debug the node.
# debug ID format is `container_name.pod_name.namespace`.
$ sealer debug node mynode

# Debug node `mynode` with envionment variables.
$ sealer debug node mynode -it --env yyds=sealer,debug=sealer

# Debug node `mynode` and set some check items. After attaching to the debug pod , it will check the specified check items automatic and show the check results.
# Attention: currently, `sealer debug` can only use this with default image.
$ sealer debug node mynode -it --check-list=dns,network

# Debug node `mynode`, and we can specify the container name in debug pod.
$ sealer debug node mynode -it --name sealer-demo
```

Just like `sealer debug node`,  so only list  some case

```bash
# Debug pod `mypod` and immediately attach to the debug container.
$ sealer debug pod mypod -it

# Debug pod `mypod`and we can specify the debug container name.
$ sealer debug pod mypod -it --name sealer-demo

# Default it will debug the first container in pod `mypod`. This way, it will debug the container `mypod` in pod `mypod`.
$ sealer debug pod mypod -c podaa -it

# Debug `coredns` pod in namespace kube-system.
$ sealer debug pod coredns-b55nb -n kube-system -it
```
The `sealer debug` provides some default images.

```bash
# Show the defautl images provided by sealer.
$ sealer debug show-images
There are several default images you can use：
RichToolsOnUbuntu:  debug:ubuntu
RichToolsOnAlpine:  debug:apline
```

`sealer debug node mynode/pod mypod`  will generate a debug pod/container, but it will not be cleaned automatically. So we can use `sealer debug clean [DEBUG_ID]` to clean the generated debug pod or container. 

```bash
# Clean generated debug pod/container using debug ID.
$ sealer debug clean sealer-pod-debugger-khwbl.mynode.default
```

### Additional context
We can't use `kubectl exec` entry the `coredns` pod.
image

There are not many useful tools for debugging.
image
image

### PR DESCRIPTION
**Pull request by DawnGuoDev**
### Describe what this PR does / why we need it
This PR implements the sealer-debug and provide some debug APIs.

### Does this pull request fix one issue?
Fixes #523 

### Describe how you did it
In issue #523, I describe what I want to do and how to implement it.

### Describe how to verify it
You can use `sealer debug` just like the example commands in issue #523. 

For example, 
image

### Special notes for reviews
There are a lot of files that have changed, but many of them are module files(in /vendor). So you can just only review the file s in /debug and /sealer/cmd/debug.go .

### CODE REVIEW

File: debug/clusterinfo/clusterinfo.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	return domain, serviceClusterIP, endpointsIPs, nil
+}
+
+func removeDuplicatesAndEmpty(ss []string) []string {
+	sort.Strings(ss)
+	ret := []string{}
+	l := len(ss)
+	for i:=0; i < l; i++ {
```
Comment (allencloud) [inline debug/clusterinfo/clusterinfo.go:86]: Are you sure you have `gofmt`ed your code? 

From my knowledge, it should be `for i := 0;` here.

@DawnGuoDev

File: debug/clusterinfo/clusterinfo.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	endpointsIPs := []string{}
+	for _, address := range endpoints.Subsets[0].Addresses {
+		endpointsIPs = append(endpointsIPs, address.IP)
+	}
+
+	return domain, serviceClusterIP, endpointsIPs, nil
+}
+
+func removeDuplicatesAndEmpty(ss []string) []string {
```
Comment (allencloud) [inline debug/clusterinfo/clusterinfo.go:82]: Please add a unit test for this function.

File: debug/clusterinfo/clusterinfo.go (hunk 1 above)
Comment (fanux) [reply debug/clusterinfo/clusterinfo.go:86]: GITHUB lint action will failed 😄

Comment (DawnGuoDev) [reply debug/clusterinfo/clusterinfo.go:86]: Thanks, I will gofmt my code, I forget it.

File: debug/clusterinfo/clusterinfo.go (hunk 2 above)
Comment (DawnGuoDev) [reply debug/clusterinfo/clusterinfo.go:82]: Got it

File: debug/images.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+type DebugImagesManagement interface {
+	ShowDefaultImages() error
+	GetDefaultImage() (string, error)
+}
+const SealerRegistryUrl = "registry.cn-qingdao.aliyuncs.com/sealer-apps/"
```
Comment (allencloud) [inline debug/images.go:27]: s/SealerRegistryUrl/DefaultSealerRegistryUrl/g

Comment (DawnGuoDev) [reply debug/images.go:27]: Got it. With the help of golint, I modify this and other go code format.

**Review by fanux — APPROVED** [review]

File: debug/images.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+type ImagesManagement interface {
+	ShowDefaultImages() error
+	GetDefaultImage() (string, error)
+}
+
+const DefaultSealerRegistryURL = "registry.cn-qingdao.aliyuncs.com/sealer-apps/"
+
+var ShowImagesCMD = &cobra.Command{
+	Use:   "show-images",
```
Comment (allencloud) [inline debug/images.go:31]: I am wondering if  just `images` is OK, rather than `show-images`.

File: debug/clusterinfo/clusterinfo.go (hunk 2 above)
Comment (DawnGuoDev) [reply debug/clusterinfo/clusterinfo.go:82]: Thanks. From the time complexity, using map to remove repetition is O(n), and I will think about using it. But you say the results seems not true, can you give some examples？The result of your given example is true.

File: debug/images.go (hunk 4 above)
Comment (DawnGuoDev) [reply debug/images.go:31]: From the semantic, I think the `sealer debug show-images` are clearer to express what the command do. When using `sealer debug images`, it looks more like debugging the images. So, I used images in the beginning, but then I changed it to show-images.

File: debug/clusterinfo/clusterinfo.go (hunk 2 above)
Comment (oldthreefeng) [reply debug/clusterinfo/clusterinfo.go:82]: I'm not sure you sorts the silce.... may the result need to be sorted ?! 

when add `sort.Strings` method to `removeRep`.   perfrom is bad...

```
func removeRep(a []string) []string {
	sort.Strings(a)
	if len(a) == 0 {
		return a
	}
	res := make([]string, 0, len(a))
	tmp := map[string]struct{}{}
	for _, v := range a {
		if _, ok := tmp[v]; !ok && len(v) != 0 {
			tmp[v] = struct{}{}
			res = append(res, v)
		}
	}
	return res
}

then test  Benchmark_removeRep

pkg: github.com/fanux/sealos/k8s
cpu: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
Benchmark_removeRep
Benchmark_removeRep-8   	 2358082	       512.8 ns/op	     184 B/op	       2 allocs/op
PASS
ok  	github.com/fanux/sealos/k8s	1.752s

pkg: github.com/fanux/sealos/k8s
cpu: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
Benchmark_remove01
Benchmark_remove01-8   	 2418132	       487.6 ns/op	     136 B/op	       4 allocs/op
PASS
ok  	github.com/fanux/sealos/k8s	1.708s
```

Comment (DawnGuoDev) [reply debug/clusterinfo/clusterinfo.go:82]: The results do not need to be sorted. The sorted in begin was for removing repetition.

Would it be better to change `tmp := map[string]struct{}{}` to `tmp := map[string]bool`?

Comment (oldthreefeng) [reply debug/clusterinfo/clusterinfo.go:82]: > Would it be better to change tmp := map[string]struct{}{} to tmp := map[string]bool?

yes. here you go.

**Pull request merged**